### ISSUE TITLE
Bad interaction between ambivalent types and subtyping coercion

### ISSUE THREAD

**Issue by let-def** [ocaml/ocaml#10822]
Compiling this file (tested on current trunk):

```ocaml
type t
type u = private t

type ('a, 'b) eq = Refl : ('a, 'a) eq

let foo (type s) x (Refl : (s, u) eq) =
  (x : s :> t)
```

Produces the error:

```
File "ambivalent_subtyping.ml", line 7, characters 3-4:
7 |   (x : s :> t)
       ^
Error: This expression has type 'a but an expression was expected of type
         s = u 
       This instance of u is ambiguous:
       it would escape the scope of its equation
```

This shouldn't be the case given that an explicit type is provided in the coercion (and the `'a` leaking in the message is a bit dubious). Indeed, changing the last line to `((x : s) :> t)` fixes the problem.

**Comment by let-def** [comment]
It seems to work fine with OCaml 4.12, the problem appeared in 4.13.

**Comment by garrigue** [comment]
Fixed by #10823

**Comment by Octachron** [comment]
(I have cherry-picked the fix to trunk too.)
